src/internal/auth/handler.go

```diff
@@ -1,51 +1,67 @@
 package auth
 
 import (
+	"fmt"
 	"github.com/pachyderm/pachyderm/v2/src/auth"
 	"github.com/pachyderm/pachyderm/v2/src/client"
 )
 
-type authHandler func(*client.APIClient, string) error
+// an authHandler can optionally return a username string that will be cached in the request's context
+type authHandler func(*client.APIClient, string) (string, error)
+
+type ContextKey string
+
+const WhoAmIResultKey = ContextKey("WhoAmI")
```

Comment:
I think it makes sense to make these package-private, and expose a "getWhoAmI" method that accepts a context and returns the username if it's set? That way we know it can only be set in this package.